libgstd/libgstd.c

```diff
@@ -0,0 +1,202 @@
+    GstDManager ** out, GOptionGroup ** gst_group, int argc, char *argv[])
+{
+  GstdStatus ret = GSTD_LIB_OK;
+  GstDManager *manager = NULL;
+  GstdSession *session = NULL;
+  GstdIpc **ipc_array = NULL;
+
+  gstd_assert_and_ret_val (NULL != out, GSTD_NULL_ARGUMENT);
+
+  manager = (GstDManager *) g_malloc0 (sizeof (*manager));
+  session = gstd_session_new ("Session0");
+
+  /* If there is ipcs, then initialize them */
+  if (NULL != supported_ipcs && num_ipcs > 0) {
+    ipc_array = g_malloc0 (num_ipcs * sizeof (*ipc_array));
+    for (guint i = 0; i < num_ipcs; i++) {
+      ipc_array[i] =
+          GSTD_IPC (g_object_new (gstd_supported_ipc_to_ipc (supported_ipcs[i]),
+              NULL));
+    }
+    manager->ipc_array = ipc_array;
+  }
+
+  manager->session = session;
+  manager->num_ipcs = num_ipcs;
+
+  *out = manager;
+
+  /* Initialize GStreamer */
+  gstd_manager_init (gst_group, argc, argv);
+
+  return ret;
+}
+
+void
+gstd_manager_ipc_options (GstDManager * manager, GOptionGroup * ipc_group[])
+{
+  gint i = 0;
+
+  gstd_assert_and_ret (NULL != manager);
+  gstd_assert_and_ret (NULL != manager->ipc_array);
+  gstd_assert_and_ret (NULL != ipc_group);
+
+  for (i = 0; i < manager->num_ipcs; i++) {
+    gstd_ipc_get_option_group (manager->ipc_array[i], &ipc_group[i]);
+  }
+}
+
+gboolean
+gstd_manager_ipc_start (GstDManager * manager)
+{
+  gboolean ipc_selected = FALSE;
+  gboolean ret = TRUE;
+  GstdReturnCode code = GSTD_EOK;
+  gint i = 0;
+
+  gstd_assert_and_ret_val (NULL != manager, GSTD_NULL_ARGUMENT);
+  gstd_assert_and_ret_val (NULL != manager->ipc_array, GSTD_LIB_NOT_FOUND);
+  gstd_assert_and_ret_val (NULL != manager->session, GSTD_LIB_NOT_FOUND);
+
+  /* Verify if at least one IPC mechanism was selected */
+  for (i = 0; i < manager->num_ipcs; i++) {
+    g_object_get (G_OBJECT (manager->ipc_array[i]), "enabled", &ipc_selected,
+        NULL);
+
+    if (ipc_selected) {
+      break;
+    }
+  }
+
+  /* If no IPC was selected, default to TCP */
+  if (!ipc_selected) {
+    g_object_set (G_OBJECT (manager->ipc_array[0]), "enabled", TRUE, NULL);
+  }
+
+  /* Run start for each IPC (each start method checks for the enabled flag) */
+  for (i = 0; i < manager->num_ipcs; i++) {
+    code = gstd_ipc_start (manager->ipc_array[i], manager->session);
+    if (code) {
+      g_printerr ("Couldn't start IPC : (%s)\n",
+          G_OBJECT_TYPE_NAME (manager->ipc_array[i]));
+      ret = FALSE;
+    }
+  }
+
+  return ret;
+}
+
+void
+gstd_manager_ipc_stop (GstDManager * manager)
+{
+  gint i = 0;
+
+  gstd_assert_and_ret (NULL != manager);
+  gstd_assert_and_ret (NULL != manager->ipc_array);
+  gstd_assert_and_ret (NULL != manager->session);
+
+  /* Run stop for each IPC */
+  for (i = 0; i < manager->num_ipcs; i++) {
+    if (NULL != manager->ipc_array[i] && TRUE == manager->ipc_array[i]->enabled) {
+      gstd_ipc_stop (manager->ipc_array[i]);
+      g_clear_object (&manager->ipc_array[i]);
+    }
```

Comment:
I agree that we shouldn't stop if it isn't enabled, but we should always call `g_object_clear`, right?

Reply:
Doing that you get the error `GLib-CRITICAL **:  g_thread_pool_free: assertion 'real' failed`

But there is not problem no calling `g_object_clear` because it decrease the count reference of the object and set the pointer to NULL. So if it's already NULL, you don't need to do anything with it